pytorch_lightning/trainer/training_loop.py

```diff
@@ -515,21 +521,30 @@ def run_training_epoch(self):
             # ------------------------------------
             batch_output = self.run_training_batch(batch, batch_idx, dataloader_idx)
 
-            # when returning -1 from train_step, we end epoch early
-            if batch_output.signal == -1:
-                break
+            if batch_output is not None:
+                # only track outputs when user implements training_epoch_end
+                # otherwise we will build up unnecessary memory
+                epoch_end_outputs = self.process_train_step_outputs(
+                    batch_output.training_step_output_for_epoch_end,
+                    self.early_stopping_accumulator,
+                    self.checkpoint_accumulator
+                )
 
-            # only track outputs when user implements training_epoch_end
-            # otherwise we will build up unnecessary memory
-            epoch_end_outputs = self.process_train_step_outputs(
-                batch_output.training_step_output_for_epoch_end,
-                self.early_stopping_accumulator,
-                self.checkpoint_accumulator
-            )
+                # when returning -1 from train_step, we end epoch early
+                if batch_output.signal == -1:
+                    break
+
```

Comment:
TIL about this. where is this documented? Does it needed to be wrapped in batch_output.signal? does someone need to return `{"signal": -1}` from `training_step` to trigger this

Reply:
This is documented in Model Hooks and in the api docs https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.core.hooks.html#pytorch_lightning.core.hooks.ModelHooks.on_train_batch_start
It just need to return -1